Arc support broken

Hey, thank you for forking and taking your time to merge the fixes!

Judging the commit e84fe764003a924bd5c04d4dd2f1e0accc3c3654 I assume that you've decided to only keep the arc version of `im` and ditch `im-rc`.

So I am surprised to see, that the following code

```
use std::sync::Mutex;
use std::thread;

extern crate imbl;

fn main() {
    let x : Mutex<imbl::Vector<u32>> = Mutex::new(imbl::Vector::new());

    thread::spawn(move || {
      let _ = x;
    });
}
```

Compiles fine with the commit right before removing the Rc implementation

```
[dependencies]
imbl = {git="https://github.com/jneem/imbl.git", rev="67ce8b5976941fbf2d1c96f5c68e868a497237d8"}
```

But doesn't compile after removing the Rc implementation
```
[dependencies]
imbl = {git="https://github.com/jneem/imbl.git", rev="e84fe764003a924bd5c04d4dd2f1e0accc3c3654"}
```

```
error[E0277]: `Rc<sized_chunks::sized_chunk::Chunk<usize>>` cannot be sent between threads safely
   --> src/main.rs:10:5
    |
10  |     thread::spawn(move || {
    |     ^^^^^^^^^^^^^ `Rc<sized_chunks::sized_chunk::Chunk<usize>>` cannot be sent between threads safely
    |
    = help: within `Vector<u32>`, the trait `Send` is not implemented for `Rc<sized_chunks::sized_chunk::Chunk<usize>>`
```

I am using `rustc 1.54.0 (a178d0322 2021-07-26)` on Ubuntu 21.4

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -53,3 +53,4 @@ rand = { version = "0.8", features = ["small_rng"] }
 pretty_assertions = "0.7"
 metrohash = "1"
 proptest-derive = "0.3"
+static_assertions = "1.1.0"

--- a/src/hash/map.rs
+++ b/src/hash/map.rs
@@ -2123,7 +2123,12 @@ mod test {
     use crate::test::LolHasher;
     use ::proptest::num::{i16, usize};
     use ::proptest::{collection, proptest};
+    use static_assertions::{assert_impl_all, assert_not_impl_any};
     use std::hash::BuildHasherDefault;
+
+    assert_impl_all!(HashMap<i32, i32>: Send, Sync);
+    assert_not_impl_any!(HashMap<i32, *const i32>: Send, Sync);
+    assert_not_impl_any!(HashMap<*const i32, i32>: Send, Sync);
 
     #[test]
     fn safe_mutation() {

--- a/src/hash/set.rs
+++ b/src/hash/set.rs
@@ -1068,7 +1068,11 @@ mod test {
     use crate::test::LolHasher;
     use ::proptest::num::i16;
     use ::proptest::proptest;
+    use static_assertions::{assert_impl_all, assert_not_impl_any};
     use std::hash::BuildHasherDefault;
+
+    assert_impl_all!(HashSet<i32>: Send, Sync);
+    assert_not_impl_any!(HashSet<*const i32>: Send, Sync);
 
     #[test]
     fn insert_failing() {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -54,7 +54,7 @@
 //! just pick the one that has the right shape and it should have
 //! acceptable performance characteristics for every operation you
 //! might need. Specialised data structures will always be faster at
-//! what they've been specialised for, but `im` aims to provide the
+//! what they've been specialised for, but `imbl` aims to provide the
 //! data structures which deliver the least chance of accidentally
 //! using them for the wrong thing.
 //!
@@ -282,33 +282,28 @@
 //!
 //! ## Thread Safety
 //!
-//! The data structures in the `im` crate are thread safe, through
-//! [`Arc`][std::sync::Arc]. This comes with a slight performance impact, so
-//! that if you prioritise speed over thread safety, you may want to use the
-//! `im-rc` crate instead, which is identical to `im` except that it uses
-//! [`Rc`][std::rc::Rc] instead of [`Arc`][std::sync::Arc], implying that the
-//! data structures in `im-rc` do not implement [`Send`][std::marker::Send] and
-//! [`Sync`][std::marker::Sync]. This yields approximately a 20-25% increase in
-//! general performance.
+//! The data structures in the `imbl` crate are thread safe, through [`Arc`][std::sync::Arc]. This
+//! comes with a slight performance impact. The original `im` crate had a non-threadsafe version
+//! called `im-rc`. If you prioritise speed over thread safety, you may want to use that crate
+//! instead.  We intend to eventually support a faster, non-threadsafe version of `imbl`.
 //!
 //! ## Feature Flags
 //!
-//! `im` comes with optional support for the following crates through Cargo
+//! `imbl` comes with optional support for the following crates through Cargo
 //! feature flags. You can enable them in your `Cargo.toml` file like this:
 //!
 //! ```no_compile
 //! [dependencies]
-//! im = { version = "*", features = ["proptest", "serde"] }
+//! imbl = { version = "*", features = ["proptest", "serde"] }
 //! ```
 //!
 //! | Feature | Description |
 //! | ------- | ----------- |
-//! | [`pool`](https://crates.io/crates/refpool) | Constructors and pool types for [`refpool`](https://crates.io/crates/refpool) memory pools (only available in `im-rc`) |
-//! | [`proptest`](https://crates.io/crates/proptest) | Strategies for all `im` datatypes under a `proptest` namespace, eg. `imbl::vector::proptest::vector()` |
-//! | [`quickcheck`](https://crates.io/crates/quickcheck) | [`quickcheck::Arbitrary`](https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html) implementations for all `im` datatypes (not available in `im-rc`) |
-//! | [`rayon`](https://crates.io/crates/rayon) | parallel iterator implementations for [`Vector`][vector::Vector] (not available in `im-rc`) |
-//! | [`serde`](https://crates.io/crates/serde) | [`Serialize`](https://docs.rs/serde/latest/serde/trait.Serialize.html) and [`Deserialize`](https://docs.rs/serde/latest/serde/trait.Deserialize.html) implementations for all `im` datatypes |
-//! | [`arbitrary`](https://crates.io/crates/arbitrary/) | [`arbitrary::Arbitrary`](https://docs.rs/arbitrary/latest/arbitrary/trait.Arbitrary.html) implementations for all `im` datatypes |
+//! | [`proptest`](https://crates.io/crates/proptest) | Strategies for all `imbl` datatypes under a `proptest` namespace, eg. `imbl::vector::proptest::vector()` |
+//! | [`quickcheck`](https://crates.io/crates/quickcheck) | [`quickcheck::Arbitrary`](https://docs.rs/quickcheck/latest/quickcheck/trait.Arbitrary.html) implementations for all `imbl` datatypes |
+//! | [`rayon`](https://crates.io/crates/rayon) | parallel iterator implementations for [`Vector`][vector::Vector] |
+//! | [`serde`](https://crates.io/crates/serde) | [`Serialize`](https://docs.rs/serde/latest/serde/trait.Serialize.html) and [`Deserialize`](https://docs.rs/serde/latest/serde/trait.Deserialize.html) implementations for all `imbl` datatypes |
+//! | [`arbitrary`](https://crates.io/crates/arbitrary/) | [`arbitrary::Arbitrary`](https://docs.rs/arbitrary/latest/arbitrary/trait.Arbitrary.html) implementations for all `imbl` datatypes |
 //!
 //! [std::collections]: https://doc.rust-lang.org/std/collections/index.html
 //! [std::collections::VecDeque]: https://doc.rust-lang.org/std/collections/struct.VecDeque.html
@@ -378,16 +373,16 @@ pub mod ser;
 #[doc(hidden)]
 pub mod arbitrary;
 
-#[cfg(all(threadsafe, feature = "quickcheck"))]
+#[cfg(feature = "quickcheck")]
 #[doc(hidden)]
 pub mod quickcheck;
 
-#[cfg(any(threadsafe, not(feature = "pool")))]
+#[cfg(not(feature = "pool"))]
 mod fakepool;
 
-#[cfg(all(threadsafe, feature = "pool"))]
+#[cfg(feature = "pool")]
 compile_error!(
-    "The `pool` feature is not threadsafe but you've enabled it on a threadsafe version of `im`."
+    "The `pool` feature is not threadsafe but you've enabled it on a threadsafe version of `imbl`."
 );
 
 pub use crate::hashmap::HashMap;

--- a/src/ord/map.rs
+++ b/src/ord/map.rs
@@ -2195,6 +2195,11 @@ mod test {
     use crate::test::is_sorted;
     use ::proptest::num::{i16, usize};
     use ::proptest::{bool, collection, proptest};
+    use static_assertions::{assert_impl_all, assert_not_impl_any};
+
+    assert_impl_all!(OrdMap<i32, i32>: Send, Sync);
+    assert_not_impl_any!(OrdMap<i32, *const i32>: Send, Sync);
+    assert_not_impl_any!(OrdMap<*const i32, i32>: Send, Sync);
 
     #[test]
     fn iterates_in_order() {

--- a/src/ord/set.rs
+++ b/src/ord/set.rs
@@ -1180,6 +1180,10 @@ mod test {
     use super::*;
     use crate::proptest::*;
     use ::proptest::proptest;
+    use static_assertions::{assert_impl_all, assert_not_impl_any};
+
+    assert_impl_all!(OrdSet<i32>: Send, Sync);
+    assert_not_impl_any!(OrdSet<*const i32>: Send, Sync);
 
     #[test]
     fn match_strings_with_string_slices() {

--- a/src/quickcheck.rs
+++ b/src/quickcheck.rs
@@ -4,19 +4,19 @@ use std::hash::{BuildHasher, Hash};
 use std::iter::FromIterator;
 
 impl<A: Arbitrary + Sync + Clone> Arbitrary for Vector<A> {
-    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+    fn arbitrary(g: &mut Gen) -> Self {
         Vector::from_iter(Vec::<A>::arbitrary(g))
     }
 }
 
 impl<K: Ord + Clone + Arbitrary + Sync, V: Clone + Arbitrary + Sync> Arbitrary for OrdMap<K, V> {
-    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+    fn arbitrary(g: &mut Gen) -> Self {
         OrdMap::from_iter(Vec::<(K, V)>::arbitrary(g))
     }
 }
 
 impl<A: Ord + Clone + Arbitrary + Sync> Arbitrary for OrdSet<A> {
-    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+    fn arbitrary(g: &mut Gen) -> Self {
         OrdSet::from_iter(Vec::<A>::arbitrary(g))
     }
 }
@@ -26,7 +26,7 @@ where
     A: Hash + Eq + Arbitrary + Sync,
     S: BuildHasher + Default + Send + Sync + 'static,
 {
-    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+    fn arbitrary(g: &mut Gen) -> Self {
         HashSet::from_iter(Vec::<A>::arbitrary(g))
     }
 }
@@ -37,7 +37,7 @@ where
     V: Arbitrary + Sync,
     S: BuildHasher + Default + Send + Sync + 'static,
 {
-    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+    fn arbitrary(g: &mut Gen) -> Self {
         HashMap::from(Vec::<(K, V)>::arbitrary(g))
     }
 }

--- a/src/sync.rs
+++ b/src/sync.rs
@@ -4,7 +4,8 @@
 
 pub(crate) use self::lock::Lock;
 
-#[cfg(threadsafe)]
+// FIXME: This is intended to be replaced by something else when using the non-threadsafe, `Rc`
+// version of this crate. We have temporarily disabled that.
 mod lock {
     use std::sync::{Arc, Mutex, MutexGuard};
 
@@ -23,39 +24,6 @@ mod lock {
         #[inline]
         pub(crate) fn lock(&mut self) -> Option<MutexGuard<'_, A>> {
             self.lock.lock().ok()
-        }
-    }
-
-    impl<A> Clone for Lock<A> {
-        fn clone(&self) -> Self {
-            Lock {
-                lock: self.lock.clone(),
-            }
-        }
-    }
-}
-
-#[cfg(not(threadsafe))]
-mod lock {
-    use std::cell::{RefCell, RefMut};
-    use std::rc::Rc;
-
-    /// Single threaded lock: a `RefCell` so we should safely panic if somehow
-    /// trying to access the stored data twice from the same thread.
-    pub(crate) struct Lock<A> {
-        lock: Rc<RefCell<A>>,
-    }
-
-    impl<A> Lock<A> {
-        pub(crate) fn new(value: A) -> Self {
-            Lock {
-                lock: Rc::new(RefCell::new(value)),
-            }
-        }
-
-        #[inline]
-        pub(crate) fn lock(&mut self) -> Option<RefMut<'_, A>> {
-            self.lock.try_borrow_mut().ok()
         }
     }
 

--- a/src/util.rs
+++ b/src/util.rs
@@ -12,28 +12,12 @@ use std::ptr;
 pub(crate) use refpool::{PoolClone, PoolDefault};
 
 // The `Ref` type is an alias for either `Rc` or `Arc`, user's choice.
-
+// FIXME: we have temporarily disabled `Rc`, so this is always `Arc`.
 // `Arc` without refpool
-#[cfg(all(threadsafe))]
 pub(crate) use crate::fakepool::{Arc as PoolRef, Pool, PoolClone, PoolDefault};
 
 // `Ref` == `Arc` when threadsafe
-#[cfg(threadsafe)]
 pub(crate) type Ref<A> = std::sync::Arc<A>;
-
-// `Rc` without refpool
-#[cfg(all(not(threadsafe), not(feature = "pool")))]
-pub(crate) use crate::fakepool::{Pool, PoolClone, PoolDefault, Rc as PoolRef};
-
-// `Rc` with refpool
-#[cfg(all(not(threadsafe), feature = "pool"))]
-pub(crate) type PoolRef<A> = refpool::PoolRef<A>;
-#[cfg(all(not(threadsafe), feature = "pool"))]
-pub(crate) type Pool<A> = refpool::Pool<A>;
-
-// `Ref` == `Rc` when not threadsafe
-#[cfg(not(threadsafe))]
-pub(crate) type Ref<A> = std::rc::Rc<A>;
 
 pub(crate) fn clone_ref<A>(r: Ref<A>) -> A
 where

--- a/src/vector/focus.rs
+++ b/src/vector/focus.rs
@@ -278,10 +278,8 @@ impl<A> Clone for TreeFocus<A> {
 }
 
 #[allow(unsafe_code)]
-#[cfg(threadsafe)]
 unsafe impl<A: Send> Send for TreeFocus<A> {}
 #[allow(unsafe_code)]
-#[cfg(threadsafe)]
 unsafe impl<A: Sync> Sync for TreeFocus<A> {}
 
 #[inline]

--- a/src/vector/mod.rs
+++ b/src/vector/mod.rs
@@ -68,7 +68,7 @@ pub use self::focus::{Focus, FocusMut};
 mod pool;
 pub use self::pool::RRBPool;
 
-#[cfg(all(threadsafe, any(test, feature = "rayon")))]
+#[cfg(any(test, feature = "rayon"))]
 pub mod rayon;
 
 /// Construct a vector from a sequence of elements.
@@ -2269,6 +2269,10 @@ mod test {
     use ::proptest::collection::vec;
     use ::proptest::num::{i32, usize};
     use ::proptest::proptest;
+    use static_assertions::{assert_impl_all, assert_not_impl_any};
+
+    assert_impl_all!(Vector<i32>: Send, Sync);
+    assert_not_impl_any!(Vector<*const i32>: Send, Sync);
 
     #[test]
     fn macro_allows_trailing_comma() {

